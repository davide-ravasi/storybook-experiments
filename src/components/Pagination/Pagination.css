.paging {
  display: flex;
  margin-top: 30px;
}

.paging li {
  cursor: pointer;
  border: 1px solid #ccc;
  padding: 0 6px;
}

.paging li.current {
  color: #0f0;
  border: 1px solid #0f0;
  font-weight: 700;
}

.paging li:hover {
  color: #f00;
  border: 1px solid #f00;
}