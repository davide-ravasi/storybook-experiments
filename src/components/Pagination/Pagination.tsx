import React, { useEffect, useState } from "react";
import './Pagination.css';

interface PaginationProps {
  totalItems: number;
  itemsPage: number;
}

const Pagination = ({ totalItems = 10, itemsPage = 5 }: PaginationProps) => {
  const url = 'https://randomuser.me/api?results=' + totalItems;
  const itemsPerPage = itemsPage;
  const [actualPage, setActualPage] = useState(1);
  const [content, setContent] = useState([]);

  useEffect(() => {
    const fetchList = async () => {
      let response = await fetch(url);
      if (response.ok) {
        let { results } = await response.json();
        setContent(results);
      }
    }

    fetchList();
  }, []);


  return (
    <div>
      <Pages content={content} itemsPerPage={itemsPerPage} actualPage={actualPage} />
      <Paging itemsAmount={content.length} itemsPerPage={itemsPerPage} actualPage={actualPage} setActualPage={setActualPage} />
    </div>
  )
};

interface PagingProps {
  itemsAmount: number;
  itemsPerPage: number;
  actualPage: number;
  setActualPage: React.Dispatch<React.SetStateAction<number>>;
}

const Paging = ({ itemsAmount, itemsPerPage, actualPage, setActualPage }: PagingProps) => {
  const [pages, setPages] = useState(0);

  const setNewPage = (page: number) => {
    console.log("page ", page);
    setActualPage(page);
  }

  useEffect(() => {
    setPages(Math.ceil(itemsAmount / itemsPerPage));
  }, [itemsAmount, itemsPerPage]);

  return (
    <ul className="paging">
      {pages && [...Array(pages)].map((page, index) => {
        const pageNumber = index + 1;

        return (
          <li onClick={() => setNewPage(pageNumber)} className={actualPage === pageNumber && 'current'}>{pageNumber}</li>
        )
      })}
    </ul>
  )
}


interface PageProps {
  content: any,
  itemsPerPage: number;
  actualPage: number;
}

// generic component for displaying table from array
// of objects where fields of objects are columns
const Pages = ({ content, itemsPerPage, actualPage }: PageProps) => {
  const [pageContent, setPageContent] = useState([]);

  useEffect(() => {
    if (content) {
      let newList = content.slice(actualPage, actualPage + itemsPerPage);
      setPageContent(newList);
    }
  }, [content, itemsPerPage, actualPage]);

  return (
    <table>
      <thead>
        <tr>
          <th>First Name</th>
          <th>LastName</th>
        </tr>
      </thead>
      <tbody>
        {pageContent && pageContent.map(element => {
          return (
            <tr>
              <td>{element.name.first}</td>
              <td>{element.name.last}</td>
            </tr>
          )
        })}
      </tbody>
    </table>
  )

};

export default Pagination;