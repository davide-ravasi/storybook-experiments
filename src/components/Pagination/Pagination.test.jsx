import { render, screen, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import Pagination from "./Pagination";

describe("Custom program component", () => {
  it("list elements should be presents", async () => {
    render(<Pagination totalItems="4" itemsPage="2" />);

    // check the list
    const listNames = await screen.findByRole("table", { name: "Names list:" });
    expect(listNames).toBeInTheDocument();

    // check elements in the list
    const { findAllByRole } = within(listNames);
    const items = await findAllByRole("columnheader");
    expect(items.length).toBe(2);
    const items2 = await findAllByRole("cell");
    expect(items2.length).toBe(6);
  });

  it("paging elements should be presents", async () => {
    render(<Pagination totalItems="4" itemsPage="2" />);

    // check the pagination
    const listPages = await screen.findByRole("list", {
      name: "Navigate between the pages",
    });
    expect(listPages).toBeInTheDocument();

    // check elements in pagination
    const { findAllByRole } = within(listPages);
    const listPagesItems = await findAllByRole("listitem");
    expect(listPagesItems.length).toBe(2);

    // check selected elements in pagination at start
    expect(listPagesItems[0]).toHaveClass("current");
    expect(listPagesItems[0]).toHaveTextContent("1");
  });

  it("check change pages", async () => {
    render(<Pagination totalItems="4" itemsPage="2" />);

    const listPages = await screen.findByRole("list", {
      name: "Navigate between the pages",
    });

    // check elements in pagination
    const { findAllByRole } = within(listPages);
    const listPagesItems = await findAllByRole("listitem");
    const listPageTwo = listPagesItems[1];
    expect(listPageTwo).toHaveTextContent("2");

    userEvent.click(listPageTwo);

    // check clicked element in pagination
    expect(listPageTwo).toHaveClass("current");

    const listNames = await screen.findByRole("table", { name: "Names list:" });

    // mock datas
    expect(listNames).toMatchSnapshot();
  });
});
