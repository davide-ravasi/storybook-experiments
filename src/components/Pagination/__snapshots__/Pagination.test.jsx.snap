// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom program component check change pages 1`] = `
<table>
  <caption>
    Names list:
  </caption>
  <thead>
    <tr>
      <th>
        First Name
      </th>
      <th>
        LastName
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Lucien
      </td>
      <td>
        Aubert
      </td>
    </tr>
    <tr>
      <td>
        Alisha
      </td>
      <td>
        Rodriguez
      </td>
    </tr>
  </tbody>
</table>
`;
